### ISSUE TITLE
Implicit HeaderDecorator for unsafeReadCsv is only found if asCsv is also called

### ISSUE THREAD

**Issue by seanf** [nrinaudo/kantan.csv#185]
I found that some parsing code, which worked in one file, wouldn't even compile in another file (even with the same imports). I eventually narrowed it down to the presence of a call to `List[CaseClass]().asCsv(rfc)` elsewhere in the scope of the working code.  No doubt there's some good reason for this if I fully understood implicits...

```scala
import kantan.codecs.{Decoder, Optional}
import kantan.csv._
import kantan.csv.ops._
import kantan.csv.generic._
import org.scalatest.FunSpec

class DecoderTest extends FunSpec  {

  case class CaseClass (
    num: Int,
    optionalColumn: Foo,
  )

  sealed trait Foo
  object One extends Foo
  object Two extends Foo
  object Unknown extends Foo

  private final val OneToken = "One"
  private final val TwoToken = "Two"

  implicit val fooCodec: CellCodec[Foo] = {
    CellCodec.from(s => DecodeResult(s match {
      case OneToken => One
      case TwoToken => Two
      case _ => Unknown
    })) {
      case One => OneToken
      case Two => TwoToken
      case Unknown => ""
    }
  }
  implicit def fooDecoder[E: Optional, F, T](
    implicit da: Decoder[E, Foo, F, T]): Decoder[E, Foo, F, T] =
    Decoder.from { e ⇒
      if (Optional[E].isEmpty(e)) Right(Unknown)
      else da.decode(e)
    }
//   implicit val caseClassCodec: RowCodec[CaseClass] = RowCodec.caseCodec(0, 1)(CaseClass.apply)(CaseClass.unapply)

  ignore("if this is commented out, the next test won't compile!") {
//    List[CaseClass]().asCsv(rfc)
  }

  it("from CSV to CaseClass (missing column)") {
    val csv =
      """12345
        |12345
        |""".stripMargin
    val data = csv.unsafeReadCsv[Iterable, CaseClass](rfc)
    println(data)
  }
}

```

The compile error:
```
Error:(52, 54) could not find implicit value for evidence parameter of type kantan.csv.HeaderDecoder[DecoderTest.this.CaseClass]
    val data = csv.unsafeReadCsv[Iterable, CaseClass](rfc)
Error:(52, 54) not enough arguments for method unsafeReadCsv: (implicit evidence$9: kantan.csv.HeaderDecoder[DecoderTest.this.CaseClass], implicit e: kantan.csv.engine.ReaderEngine, implicit cbf: scala.collection.generic.CanBuildFrom[Nothing,DecoderTest.this.CaseClass,Iterable[DecoderTest.this.CaseClass]])Iterable[DecoderTest.this.CaseClass].
Unspecified value parameters evidence$9, e, cbf.
    val data = csv.unsafeReadCsv[Iterable, CaseClass](rfc)
```

Related: https://github.com/nrinaudo/kantan.csv/issues/107 and https://github.com/nrinaudo/kantan.csv/issues/184

**Comment by nrinaudo** [comment]
Again, can't reproduce. I just copied this in the generic's module test code, and executed tests both with and without the `List[CaseClass].asCsv(rfc)` line commented out, and bot cases worked. I suspect there's something *else* you're doing and not realising.

**Comment by seanf** [comment]
Thanks @nrinaudo.

This stuff is too weird...

I just changed my `DecoderTest` class (as above) to `object DecoderTest extends App`, removed the lines which mentioned ScalaTest's `ignore()` and `it()`, and now the code compiles, with or without `List[CaseClass]().asCsv(rfc)`. 

You said it worked for you, but after that change, I found it fails with `ArrayIndexOutOfBoundsException` as in #184. I tried adding an ordered `RowDecoder`:

```scala
implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)
```

but it won't compile:
```
Error:(40, 67) diverging implicit expansion for type kantan.csv.CellDecoder[DecoderTest.Foo]
starting with method decoderFromExported in object Decoder
  implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)
Error:(40, 67) not enough arguments for method ordered: (implicit evidence$255: kantan.csv.CellDecoder[Int], implicit evidence$256: kantan.csv.CellDecoder[DecoderTest.Foo])kantan.csv.RowDecoder[DecoderTest.CaseClass].
Unspecified value parameter evidence$256.
  implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)
```

Can you give me the version you tried, please?

Interestingly, if I add some extra columns to the CSV like this:

```scala
  val csv =
    """12345,One,2
      |12345,Two,3
      |""".stripMargin
```

Then `unsafeReadCsv` quietly discards the extra columns, even though there are more columns than parameters to the case class.

**Comment by nrinaudo** [comment]
It's very hard to follow what exactly is happening when you explain a diff in words - I still can't reproduce, but maybe it's because I'm not understanding what you're saying you've done, or not doing it right?

Could you please put in a minimal reproduction case for each behaviour you feel is incorrect, explain what you'd expect to happen and what actually happens?

**Comment by seanf** [comment]
Thanks @nrinaudo, let me try that again.

1. Firstly, I would expect the original FunSpec class at the very top of this issue to compile exactly as it is. Have you tried that file as is? It seems like FunSpec might be the trigger. 

The unexpected compiler errors are listed in the issue description too. These are the dependencies:

```scala
scalaVersion := "2.12.8"
libraryDependencies += "com.nrinaudo" %% "kantan.csv" % "0.5.0"
libraryDependencies += "com.nrinaudo" % "kantan.csv-generic_2.12" % "0.5.0"
libraryDependencies += "org.scalactic" %% "scalactic" % "3.0.5"
libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.5" % "test"
```

2. And here's the slightly modified `object` version which doesn't use ScalaTest:
```scala
import kantan.codecs.{Decoder, Optional}
import kantan.csv._
import kantan.csv.ops._
import kantan.csv.generic._

object DecoderTest extends App {

  case class CaseClass (
    num: Int,
    optionalColumn: Foo,
  )

  sealed trait Foo
  object One extends Foo
  object Two extends Foo
  object Unknown extends Foo

  private final val OneToken = "One"
  private final val TwoToken = "Two"

  implicit val fooCodec: CellCodec[Foo] = {
    CellCodec.from(s => DecodeResult(s match {
      case OneToken => One
      case TwoToken => Two
      case _ => Unknown
    })) {
      case One => OneToken
      case Two => TwoToken
      case Unknown => ""
    }
  }
  implicit def fooDecoder[E: Optional, F, T](
    implicit da: Decoder[E, Foo, F, T]): Decoder[E, Foo, F, T] =
    Decoder.from { e ⇒
      if (Optional[E].isEmpty(e)) Right(Unknown)
      else da.decode(e)
    }
  //   implicit val foo2Codec: RowCodec[CaseClass] = RowCodec.caseCodec(0, 1)(CaseClass.apply)(CaseClass.unapply)

  List[CaseClass]().asCsv(rfc)

  val csv =
    """12345
      |12345
      |""".stripMargin
  val data = csv.unsafeReadCsv[Iterable, CaseClass](rfc)
  println(data)

}
```

This one does compile, but at runtime I would expect it to parse both rows successfully into instances of `CaseClass`, with `Unknown` for the missing column. (Note: the CSV data only has one column, whereas the `CaseClass` has two parameters, but `implicit def fooDecoder` is intended to convert the missing column into `Unknown`.)

What actually happens: it fails with `ArrayIndexOutOfBoundsException`:

```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Array index out of range: 0
	at kantan.csv.CsvSource.$anonfun$unsafeReader$2(CsvSource.scala:96)
	at scala.util.Either$LeftProjection.map(Either.scala:573)
	at kantan.csv.CsvSource.$anonfun$unsafeReader$1(CsvSource.scala:92)
	at kantan.codecs.resource.ResourceIterator$$anon$6.readNext(ResourceIterator.scala:283)
	at kantan.codecs.resource.ResourceIterator.next(ResourceIterator.scala:114)
	at kantan.codecs.resource.ResourceIterator.next$(ResourceIterator.scala:108)
	at kantan.codecs.resource.ResourceIterator$$anon$6.next(ResourceIterator.scala:281)
	at kantan.codecs.resource.ResourceIterator.foreach(ResourceIterator.scala:318)
	at kantan.codecs.resource.ResourceIterator.foreach$(ResourceIterator.scala:318)
	at kantan.codecs.resource.ResourceIterator$$anon$6.foreach(ResourceIterator.scala:281)
	at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
	at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:662)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:639)
	at scala.collection.TraversableOnce.to(TraversableOnce.scala:313)
	at scala.collection.TraversableOnce.to$(TraversableOnce.scala:311)
	at kantan.codecs.resource.ResourceIterator$$anon$6.to(ResourceIterator.scala:281)
	at kantan.csv.CsvSource.unsafeRead(CsvSource.scala:153)
	at kantan.csv.CsvSource.unsafeRead$(CsvSource.scala:151)
	at kantan.csv.CsvSource$$anon$1.unsafeRead(CsvSource.scala:232)
	at kantan.csv.ops.CsvSourceOps.unsafeReadCsv(CsvSourceOps.scala:161)
	at DecoderTest$.delayedEndpoint$DecoderTest$1(DecoderTest.scala:46)
	at DecoderTest$delayedInit$body.apply(DecoderTest.scala:6)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
	at scala.App.$anonfun$main$1$adapted(App.scala:80)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.App.main(App.scala:80)
	at scala.App.main$(App.scala:78)
	at DecoderTest$.main(DecoderTest.scala:6)
	at DecoderTest.main(DecoderTest.scala)
```

3. Perhaps my expectations for case 2 are wrong, and you wouldn't expect that to work, so I tried another version which just adds an `implicit val` with `RowDecoder.ordered`:

```scala
import kantan.codecs.{Decoder, Optional}
import kantan.csv._
import kantan.csv.ops._
import kantan.csv.generic._

object DecoderTest extends App {

  case class CaseClass (
    num: Int,
    optionalColumn: Foo,
  )

  sealed trait Foo
  object One extends Foo
  object Two extends Foo
  object Unknown extends Foo

  private final val OneToken = "One"
  private final val TwoToken = "Two"

  implicit val fooCodec: CellCodec[Foo] = {
    CellCodec.from(s => DecodeResult(s match {
      case OneToken => One
      case TwoToken => Two
      case _ => Unknown
    })) {
      case One => OneToken
      case Two => TwoToken
      case Unknown => ""
    }
  }
  implicit def fooDecoder[E: Optional, F, T](
    implicit da: Decoder[E, Foo, F, T]): Decoder[E, Foo, F, T] =
    Decoder.from { e ⇒
      if (Optional[E].isEmpty(e)) Right(Unknown)
      else da.decode(e)
    }
  //   implicit val foo2Codec: RowCodec[CaseClass] = RowCodec.caseCodec(0, 1)(CaseClass.apply)(CaseClass.unapply)

  implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)


  List[CaseClass]().asCsv(rfc)

  val csv =
    """12345
      |12345
      |""".stripMargin
  val data = csv.unsafeReadCsv[Iterable, CaseClass](rfc)
  println(data)

}
```


Again, I was hoping it would parse the CSV into two CaseClass instances, but it fails to compile, reporting these errors:

```
Error:(40, 67) diverging implicit expansion for type kantan.csv.CellDecoder[DecoderTest.Foo]
starting with method decoderFromExported in object Decoder
  implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)
Error:(40, 67) not enough arguments for method ordered: (implicit evidence$255: kantan.csv.CellDecoder[Int], implicit evidence$256: kantan.csv.CellDecoder[DecoderTest.Foo])kantan.csv.RowDecoder[DecoderTest.CaseClass].
Unspecified value parameter evidence$256.
  implicit val decoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)
```

That's probably just bad code because I don't use implicits correctly, but it seems that IntelliJ can't cope with Shapeless and isn't much help here (even valid kantan.csv code shows errors in the IDE). Any help appreciated.

4. Finally, in my attempts to create a public minimal reproducer for the code in case 2, I happened to leave two many columns in the `val csv`, and I noticed that there was no runtime error. I would expect some sort of runtime validation error telling me that the CSV data didn't match the target, since the target CaseClass had fewer properties than the number of columns in the row.

It would still help to see the _your attempt_ at a reproducer, if possible, please. Especially if it parsed the CSV without a runtime error, as I understand you to mean.

Thanks.

**Comment by nrinaudo** [comment]
There's a lot to unpack here, I'll try to deal with all elements. Note that while reading the documentation would not have helped you with *everything*, it would have helped you with at least parts of this.

1. Runtime validation. I'm not sure what you mean, but I get the impression that your problem is that the code throws. I agree that it's bad, but you're using `unsafeReadCsv`, which *will* throw on error - that's it's purpose. If you want safe code, don't use the unsafe method, use `readCsv` instead.

That'll still result in an error, but one that's properly wrapped in the functions' return type.

2. I'm not sure what you're expecting to happen when you declare two `Foo` decoders (`fooCodec` and `fooDecoder` are both instances of `CellDecoder[Foo]`), but that's a sure way to break things in ways that you won't expect. In this instance, `fooDecoder` is ignored because its type is less specific than `fooCodec`, which I guess is not what you're expecting to happen.

It's generally better practice to declare an `Encoder` and a `Decoder` rather than a codec when you want to do something a bit fancy in one of them.

3. `CellDecoder[Foo]`. This is the idiomatic way of declaring it:
```scala
  implicit val fooDecoder: CellDecoder[Foo] = CellDecoder[Option[String]].map {
    case Some("One") => One
    case Some("Two") => Two
    case _ => Unknown
  }
```

I'm not expecting you to come up with this without more familiarity with the way kantan.csv works, this is meant as an example of how to do what you're trying to achieve. Digging in the innards of kantan.codecs, fishing out `Optional` and trying to work with that is going to be *painful*, that's meant to be internal and not at all for public consumption.

4. All that being said, it doesn't fix your problem. *That* comes from the fact that you're using the generics module, which is almost always a bad idea. In your case, it's a bad idea because it's not great at working with invalid CSV rows - and a row missing values is invalid, as far as it's concerned. Declare your own `RowDecoder[CaseClass]` and it'll work out:

```scala
import kantan.codecs.{Decoder, Optional}
import kantan.csv._
import kantan.csv.ops._
import kantan.csv.generic._

object DecoderTest extends App {

  case class CaseClass (
    num: Int,
    optionalColumn: Foo,
  )

  sealed trait Foo
  case object One extends Foo
  case object Two extends Foo
  case object Unknown extends Foo

  implicit val fooDecoder: CellDecoder[Foo] = CellDecoder[Option[String]].map {
    case Some("One") => One
    case Some("Two") => Two
    case _ => Unknown
  }

  implicit val caseClassDecoder: RowDecoder[CaseClass] = RowDecoder.ordered(CaseClass.apply _)

  val csv =
    """12345
      |12345
      |""".stripMargin
  val data = csv.readCsv[Seq, CaseClass](rfc)
  println(data)

}
```

**Comment by seanf** [comment]
Thanks @nrinaudo 

I did read most of the documentation pages at least to some extent (and a few of the issues), or I wouldn't have got this far! That's not to say that I understood it all as well as I would like.

1. No, I certainly expected an exception if it couldn't parse. (From what I saw, an important difference with the safe version (other than FP safety) is that you get a Left(error message) for each row which is unparseable, but no stack traces to help with diagnosis. For example: `Vector(Left(OutOfBounds: 0 is not a valid index), Left(OutOfBounds: 0 is not a valid index))`) With case 1, I just expected my code to compile (or not) regardless of context. Must be something about FunSpec...

When it does compile, and then fails at runtime, it would be nice to have a more helpful error message if possible (maybe something about the size of the row compared to the target type - as a user it's not clear with the 0 index came from), but that's more of a feature request than a bug report.

2 and 3. I see what you mean about having a Codec and a Decoder for the same type. Thanks for the idiomatic solution.

4. Thanks for the complete solution. What's the difference between `RowDecoder.ordered` and the default/generic decoder for a case class? Is there any reason `RowDecoder.ordered` couldn't be the default? 

5. If I define the case class with a default like this:

```scala
  case class CaseClass (
    num: Int,
    optionalColumn: Foo = Unknown)
```
Would it be possible to take advantage of that default value, so that a missing column will use the default value without repeating the default in the CellDecoder logic?

**Comment by nrinaudo** [comment]
Tuple instances are not really derived, not in the same sense as they are for case classes. See https://github.com/nrinaudo/kantan.csv/blob/master/core/shared/src/main/boilerplate/kantan/csv/TupleInstances.scala.template, tuple instances are kind of hard-coded and work within the normal typeclass instance framework.

Case classes, or ADTs in general, use shapeless, which, at least in kantan.csv's implementation, are a lot more strict. Not because I hate my users, but because I couldn't figure out how to make them less strict.

And no, there is no way that I know to use default values. It does not mean that it's impossible - I have never looked into them, nor do I really intend to, as I consider them a code smell best avoided except in some very specific circumstances.

**Comment by seanf** [comment]
Okay, thanks @nrinaudo. I'll try to keep away from the `generics` module. It really confuses IntelliJ anyway, so it's nice to get rid of all those red lines.

I'm not 100% sure what caused the original weirdness with compilation when calling `asCsv` (or not calling it), but I think I have all the right pieces to make it work now: exactly one `RowDecoder[CaseClass]` and one `CellDecoder[Foo]`. Thanks!

One more question, if you don't mind: how do you feel about this as a simplification of your `fooDecoder`? 

```scala
  implicit val fooDecoder: CellDecoder[Foo] = 
    s => DecodeResult(s match {
      case "One" => One
      case "Two" => Two
      case "" => Unknown
    })
```

In my experiments, absent columns can be processed either as `None` or as the empty string `""`, depending on the `CellDecoder`. I'm not sure if that's likely to hold true, long-term, or if it's just an artefact of the current implementation.

**Comment by nrinaudo** [comment]
There's one major flaw in that implementation: it'll fail if the cell contains something that is not `One`, `Two` or the empty string. Depending on what you want:
- replace `case ""` by `case _`, if you're ok with anything not `One` or `Two` being `Unknown`
- use `emap` otherwise. That's what I should have done in my implementation, which is incorrect as well.

The empty cell is decoded as:
- the empty string if you ask for a string
- `None` if you ask for an `Option`

There's no magic nor surprises here, so you could go with either. I think my preferred implementation would be :

```scala
CellDecoder[String].emap {
  case "One" => Right(One)
  case "Two" => Right(Two)
  case "" => Right(Unknown)
  case s => Left(s"Not a valid Foo: $s")
}
```

**Comment by seanf** [comment]
@nrinaudo Assuming I'm using the safe version, won't a MatchError translate to a Left anyway? But perhaps you consider that to be a code smell (since an exception is thrown in the process).

Could you please give me the full `implicit val` declaration for the `emap` version? I can't get it to compile.

**Comment by nrinaudo** [comment]
```scala
scala> implicit val fooDecoder: CellDecoder[Foo] = CellDecoder[String].emap {
     |   case "One" => Right(One)
     |   case "Two" => Right(Two)
     |   case "" => Right(Unknown)
     |   case s => DecodeResult.typeError(s"Not a valid Foo: $s")
     | }
fooDecoder: kantan.csv.CellDecoder[Foo] = kantan.codecs.Decoder$$anon$1@20accecf

scala> CellDecoder[Foo].decode("One")
res2: Either[kantan.csv.DecodeError,Foo] = Right(One$@47d4bffc)

scala> CellDecoder[Foo].decode("Two")
res3: Either[kantan.csv.DecodeError,Foo] = Right(Two$@421bf5ca)

scala> CellDecoder[Foo].decode("Three")
res4: Either[kantan.csv.DecodeError,Foo] = Left(TypeError: Not a valid Foo: Three)

scala> CellDecoder[Foo].decode("")
res5: Either[kantan.csv.DecodeError,Foo] = Right(Unknown$@30c87db9)
```

**Comment by nrinaudo** [comment]
And no, an exception will not translate to a `Left`. kantan.csv has methods to deal with that - `fromUnsafe`, I think - but most of the code assumes that it receives pure function as parameters and will not explicitly deal with the myriad of edge cases that can occur with impure ones.

**Comment by seanf** [comment]
I found that the `MatchError` was actually turned into a `Left(TypeError)` in this case with `readCsv`, but I take your point. Thanks for all the help @nrinaudo.
